### PR TITLE
[refactor] modify the implements of tuple & RowBatch

### PR DESCRIPTION
**Pull request by zuochunwei**
## Proposed changes

code refactor: improve code's readability, avoid const_cast

1. make loop simpler and clearer by using range-based loop grammar, it's safer than old loop style
2. iteration for _row_desc.tuple_descriptors() use index replace index and iterator mixed
3. add new function To cast_to(From from), use this union-based casting between two types to replace reinterpret_cast,  this new cast is more readable
4. avoid using the same variable name for nested loop, it's dangerous
5. add const keyword for member functions followed CppCoreGuidelines

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)
- [x ] Code refactor (Modify the code structure, format the code, etc...)
- [ ] Optimization. Including functional usability improvements and performance improvements.
- [ ] Dependency. Such as changes related to third-party components.
- [ ] Other.

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have created an issue on (Fix #ISSUE) and described the bug/feature there in detail
- [ x] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/runtime/row_batch.cpp [hunk 1]
```diff
@@ -120,37 +120,39 @@ RowBatch::RowBatch(const RowDescriptor& row_desc, const PRowBatch& input_batch,
...
     // pointers. Tuples were serialized in the order we are deserializing them in,
     // so the first occurrence of a tuple will always have a higher offset than any tuple
     // we already converted.
     for (int i = 0; i < _num_rows; ++i) {
         TupleRow* row = get_row(i);
-        std::vector<TupleDescriptor*>::const_iterator desc = tuple_descs.begin();
-        for (int j = 0; desc != tuple_descs.end(); ++desc, ++j) {
-            if ((*desc)->string_slots().empty() && (*desc)->collection_slots().empty()) {
+        for (size_t j = 0; j < tuple_descs.size(); ++j) {
+            auto desc = tuple_descs[i];
```
Comment (morningman) [inline be/src/runtime/row_batch.cpp:141]:
```suggestion
            auto desc = tuple_descs[j];
```

Comment (zuochunwei) [reply be/src/runtime/row_batch.cpp]: OK, done

File: be/src/runtime/tuple.cpp [hunk 2]
```diff
@@ -270,8 +269,9 @@ void Tuple::materialize_exprs(TupleRow* row, const TupleDescriptor& desc,
     memset(this, 0, desc.num_null_bytes());
     // Evaluate the output_slot_exprs and place the results in the tuples.
     int mat_expr_index = 0;
+    auto& slots = desc.slots();
     for (int i = 0; i < desc.slots().size(); ++i) {
```
Comment (morningman) [inline be/src/runtime/tuple.cpp:273]:
```suggestion
    for (int i = 0; i < slots.size(); ++i) {
```

Comment (zuochunwei) [reply be/src/runtime/tuple.cpp:273]: ok

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**